### PR TITLE
Gadm clustering

### LINKED ISSUE #141: Mislabeled bus names due to overlapping shapes in build shapes (problem first noticed in #134)
In the recent pull request #134, @Hazem-IEG identified that a bus station is not properly labeled in the correct country.
This issue shall be verified.

### PR DESCRIPTION
**Pull request by hazemakhalek**
# Closes #141 .

## Changes proposed in this Pull Request
The buses are only mapped to the shapes of their countries

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [ ] Code and workflow changes are sufficiently documented.
- [ ] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [ ] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [ ] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes.

### CODE REVIEW

File: scripts/build_bus_regions.py [hunk 1]
```diff
@@ -147,28 +149,38 @@ def custom_voronoi_partition_pts(points,
 
 def get_gadm_shape(onshore_locs, gadm_shapes):
 
+
     def locate_bus(coords):
+        point = Point(Point(coords["x"], coords["y"]))
+        if country =='TN':
+            gadm_shapes_country = gadm_shapes.filter(like='TUN', axis=0)        #TODO change the naming to 2 letter
```
Comment (davide-f) [inline scripts/build_bus_regions.py]: This conversion can be done using the helper function
https://github.com/pypsa-meets-africa/pypsa-africa/blob/c994217e2c5fff891e39dde5fe6da6d27757c498/scripts/_helpers.py#L396

Same logic applies later (function get_id), I don't repeat the comment.

Though, why this if is needed?

File: scripts/build_bus_regions.py [hunk 2]
```diff
@@ -147,28 +149,38 @@ def custom_voronoi_partition_pts(points,
...
         try:
-            return gadm_shapes[gadm_shapes.contains(
-                Point(coords["x"], coords["y"]))].item()
+            return gadm_shapes[gadm_shapes.contains(point
+                )].item()
         except ValueError:
             # return 'not_found'
-            gadm_shapes[gadm_shapes.contains(Point(-9, 32))].item(
-            )  # TODO !!Fatal!! assigning not found to a random shape
+            return min(gadm_shapes_country, key=(point.distance))             # TODO
```
Comment (davide-f) [inline scripts/build_bus_regions.py]: May you elaborate the TODO slightly?

File: scripts/cluster_network.py [hunk 3]
```diff
@@ -490,6 +502,21 @@ def cluster_regions(busmaps, input=None, output=None):
     configure_logging(snakemake)
 
     n = pypsa.Network(snakemake.input.network)
+    #n.buses.at[['3773'], 'country']='BJ'
```
Comment (davide-f) [inline scripts/cluster_network.py]: May these be removed?

File: scripts/build_bus_regions.py [hunk 4]
```diff
@@ -147,28 +149,38 @@ def custom_voronoi_partition_pts(points,
...
+            return min(gadm_shapes_country, key=(point.distance))             # TODO 
 
     def get_id(coords):
+        point = Point(Point(coords["x"], coords["y"]))
+        if country =='TN':
+            gadm_shapes_country = gadm_shapes.filter(like='TUN', axis=0)   
+        else:
+            gadm_shapes_country = gadm_shapes.filter(like=country, axis=0)   
+                    
         try:
```
Comment (davide-f) [inline scripts/build_bus_regions.py:170]: May you add a comment about this try to ease the use by a third developer.
It seems to me that if the shape cannot be identified univocally, the closest shape is selected, right?

**Comment by davide-f** [comment]
Finally, check the checklist of the PR template; in particular, a change in release_notes, like "bug fixing of overlapping/clustering shapes" may be helpful; obviously change it as you please.

**Review by davide-f — CHANGES_REQUESTED** [review]
I added some comments in the text without the review
Moreover, there are some conflicts that are appearing due to the new merges unfortunately...

**Comment by pz-max** [comment]
Need to replace centroid by polylabel to set the bus in the visual center - correctly.  See issue: https://github.com/pypsa-meets-africa/pypsa-africa/issues/191

**Comment by hazemakhalek** [comment]
I adapted the code to the comments and it works well except with hydro enabled in the config file. When enabled, add_electricity crashes with an assertion error. If this is a known error that happens in the main branch anyway then the PR should be ready to merge. @davide-f

**Comment by davide-f** [comment]
What is the hydro issue you are referring to?
Unfortunately, I am not aware of it

**Comment by davide-f** [comment]
@Hazem-IEG to fix this, I guess it would be easy to use the validation_nigeria file that is in main right now, I guess you may have used it for debugging.

A question: in this PR there are some binary files, are they supposed to be here?
I fear we may add MBs to our git history that is already very large

**Comment by davide-f** [comment]
@Hazem-IEG Are you using the environment while doing the analysis?
The pre-commit is not passing.
Please, also complete the checklist of the PR description on top (quick to do)

**Comment by davide-f** [comment]
@Hazem-IEG I think I've fixed the issue, later we can merge it :)

**Review by davide-f — APPROVED** [review]

**Pull request merged**